File: CS/E4850/GalleryDragDropHelper.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Data;
using System.Windows.Documents;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using System.Windows.Navigation;
using System.Windows.Shapes;
using DevExpress.Xpf.Bars;
using DevExpress.Xpf.Core.Native;

namespace E4850 {
    class GalleryItemDragDropHelper : DependencyObject {
        static GalleryControl gallery;
        static GalleryItem dragableItem;
        static GalleryItemGroup container;
        static GalleryItemControl gic;
        static int originalIndex = -1;
        static Point clickPoint;

        public static readonly DependencyProperty EnableGalleryItemDraggingProperty = DependencyProperty.RegisterAttached("EnableGalleryItemDragging",
            typeof(bool), typeof(GalleryItemDragDropHelper), new PropertyMetadata(false, new PropertyChangedCallback(EnableGalleryItemDraggingPropertyChanged)));

        public static bool GetEnableGalleryItemDragging(DependencyObject d) {
            return (bool)d.GetValue(EnableGalleryItemDraggingProperty);
        }
        public static void SetEnableGalleryItemDragging(DependencyObject d, bool value) {
            d.SetValue(EnableGalleryItemDraggingProperty, value);
        }
        static void EnableGalleryItemDraggingPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e) {
            if (d is GalleryControl) {
                if ((bool)e.NewValue == true) {
                    gallery = (d as GalleryControl);
                    gallery.PreviewMouseLeftButtonDown += Gallery_PreviewMouseLeftButtonDown;
                    gallery.PreviewMouseLeftButtonUp += GalleryItemDragDropHelper_PreviewMouseLeftButtonUp;
                    gallery.MouseMove += gallery_MouseMove;
                    gallery.Drop += gallery_Drop;
                    gallery.DragOver += gallery_DragOver;

                } else {
                    if (gallery != null) {
                        gallery.PreviewMouseLeftButtonDown -= Gallery_PreviewMouseLeftButtonDown;
                        gallery.PreviewMouseLeftButtonUp -= GalleryItemDragDropHelper_PreviewMouseLeftButtonUp;
                        gallery.MouseMove -= gallery_MouseMove;
                        gallery.Drop -= gallery_Drop;
                        gallery.DragOver -= gallery_DragOver;
                    }
                }
            }
        }

        static void GalleryItemDragDropHelper_PreviewMouseLeftButtonUp(object sender, MouseButtonEventArgs e) {
            dragableItem = null;
        }
        
        static void gallery_DragOver(object sender, DragEventArgs e) {
            if (!e.Data.GetDataPresent(typeof(GalleryItem))) {
                e.Effects = DragDropEffects.None;
                e.Handled = true;
                return;
            }
        }

        static void gallery_Drop(object sender, DragEventArgs e) {
            e.Handled = true;
            if (dragableItem != null) {
                HitTestResult res = VisualTreeHelper.HitTest((GalleryControl)sender, e.GetPosition((GalleryControl)sender));
                if (res.VisualHit == null)
                    return;
                GalleryItemControl target = LayoutHelper.FindParentObject<GalleryItemControl>(res.VisualHit);
                if (target != null && target.Item != dragableItem) {
                    GalleryItemCollection targetCollection = target.Item.Group.Items;
                    GalleryItemCollection sourceCollection = dragableItem.Group.Items;
                    int targetIndex = targetCollection.IndexOf(target.Item);
                    if (targetCollection == sourceCollection) {
                        targetCollection.Move(originalIndex, targetIndex);
                    } else {
                        sourceCollection.Remove(dragableItem);
                        targetCollection.Insert(targetIndex, dragableItem);
                    }
                    dragableItem = null;
                }
            }
        }

        static void gallery_MouseMove(object sender, MouseEventArgs e) {
            if (e.LeftButton == MouseButtonState.Pressed && IsDraggingStart(e.GetPosition((GalleryControl)sender))) {
                DragDrop.DoDragDrop(gallery, dragableItem, DragDropEffects.Move);
            }
        }

        static void Gallery_PreviewMouseLeftButtonDown(object sender, MouseButtonEventArgs e) {
            clickPoint = e.GetPosition((GalleryControl)sender);
            HitTestResult res = VisualTreeHelper.HitTest((GalleryControl)sender, clickPoint);
            if (res.VisualHit == null)
                return;
            gic = LayoutHelper.FindParentObject<GalleryItemControl>(res.VisualHit);
            if (gic != null) {
                dragableItem = gic.Item;
                container = dragableItem.Group;
                originalIndex = container.Items.IndexOf(dragableItem);
            }
        }

        static bool IsDraggingStart(Point newPosition) {
            return (Math.Abs(clickPoint.X - newPosition.X) > 15 || Math.Abs(clickPoint.Y - newPosition.Y) > 15) && dragableItem != null;
        }
    }

}

File: VB/vb_E4850/GalleryDragDropHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Linq
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data
Imports System.Windows.Documents
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Imaging
Imports System.Windows.Navigation
Imports System.Windows.Shapes
Imports DevExpress.Xpf.Bars
Imports DevExpress.Xpf.Core.Native

Namespace vb_E4850
	Friend Class GalleryItemDragDropHelper
		Inherits DependencyObject
		Private Shared gallery As GalleryControl
		Private Shared dragableItem As GalleryItem
		Private Shared container As GalleryItemGroup
		Private Shared gic As GalleryItemControl
		Private Shared originalIndex As Integer = -1
		Private Shared clickPoint As Point


		Public Shared ReadOnly EnableGalleryItemDraggingProperty As DependencyProperty = DependencyProperty.RegisterAttached("EnableGalleryItemDragging", GetType(Boolean), GetType(GalleryItemDragDropHelper), New PropertyMetadata(False, New PropertyChangedCallback(AddressOf EnableGalleryItemDraggingPropertyChanged)))

		Public Shared Function GetEnableGalleryItemDragging(ByVal d As DependencyObject) As Boolean
			Return CBool(d.GetValue(EnableGalleryItemDraggingProperty))
		End Function
		Public Shared Sub SetEnableGalleryItemDragging(ByVal d As DependencyObject, ByVal value As Boolean)
			d.SetValue(EnableGalleryItemDraggingProperty, value)
		End Sub
		Private Shared Sub EnableGalleryItemDraggingPropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			If TypeOf d Is GalleryControl Then
				If CBool(e.NewValue) = True Then
					gallery = (TryCast(d, GalleryControl))
					AddHandler gallery.PreviewMouseLeftButtonDown, AddressOf Gallery_PreviewMouseLeftButtonDown
					AddHandler gallery.PreviewMouseLeftButtonUp, AddressOf GalleryItemDragDropHelper_PreviewMouseLeftButtonUp
					AddHandler gallery.MouseMove, AddressOf gallery_MouseMove
					AddHandler gallery.Drop, AddressOf gallery_Drop
					AddHandler gallery.DragOver, AddressOf gallery_DragOver

				Else
					If gallery IsNot Nothing Then
						RemoveHandler gallery.PreviewMouseLeftButtonDown, AddressOf Gallery_PreviewMouseLeftButtonDown
						RemoveHandler gallery.PreviewMouseLeftButtonUp, AddressOf GalleryItemDragDropHelper_PreviewMouseLeftButtonUp
						RemoveHandler gallery.MouseMove, AddressOf gallery_MouseMove
						RemoveHandler gallery.Drop, AddressOf gallery_Drop
						RemoveHandler gallery.DragOver, AddressOf gallery_DragOver
					End If
				End If
			End If
		End Sub

		Private Shared Sub GalleryItemDragDropHelper_PreviewMouseLeftButtonUp(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
			dragableItem = Nothing
		End Sub

		Private Shared Sub gallery_DragOver(ByVal sender As Object, ByVal e As DragEventArgs)
			If (Not e.Data.GetDataPresent(GetType(GalleryItem))) Then
				e.Effects = DragDropEffects.None
				e.Handled = True
				Return
			End If
		End Sub

		Private Shared Sub gallery_Drop(ByVal sender As Object, ByVal e As DragEventArgs)
			e.Handled = True
			If dragableItem IsNot Nothing Then
				Dim res As HitTestResult = VisualTreeHelper.HitTest(CType(sender, GalleryControl), e.GetPosition(CType(sender, GalleryControl)))
				If res.VisualHit Is Nothing Then
					Return
				End If
				Dim target As GalleryItemControl = LayoutHelper.FindParentObject(Of GalleryItemControl)(res.VisualHit)
                If target IsNot Nothing AndAlso target.Item IsNot dragableItem Then
                    Dim targetCollection As GalleryItemCollection = target.Item.Group.Items
                    Dim sourceCollection As GalleryItemCollection = dragableItem.Group.Items
                    Dim targetIndex As Integer = targetCollection.IndexOf(target.Item)
                    If targetCollection Is sourceCollection Then
                        targetCollection.Move(originalIndex, targetIndex)
                    Else
                        sourceCollection.Remove(dragableItem)
                        targetCollection.Insert(targetIndex, dragableItem)
                    End If
                    dragableItem = Nothing
                End If
			End If
		End Sub

		Private Shared Sub gallery_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			If e.LeftButton = MouseButtonState.Pressed AndAlso IsDraggingStart(e.GetPosition(CType(sender, GalleryControl))) Then
				DragDrop.DoDragDrop(gallery, dragableItem, DragDropEffects.Move)
			End If
		End Sub

		Private Shared Sub Gallery_PreviewMouseLeftButtonDown(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
			clickPoint = e.GetPosition(CType(sender, GalleryControl))
			Dim res As HitTestResult = VisualTreeHelper.HitTest(CType(sender, GalleryControl), clickPoint)
			If res.VisualHit Is Nothing Then
				Return
			End If
			gic = LayoutHelper.FindParentObject(Of GalleryItemControl)(res.VisualHit)
			If gic IsNot Nothing Then
				dragableItem = gic.Item
				container = dragableItem.Group
				originalIndex = container.Items.IndexOf(dragableItem)
			End If
		End Sub

		Private Shared Function IsDraggingStart(ByVal newPosition As Point) As Boolean
			Return (Math.Abs(clickPoint.X - newPosition.X) > 15 OrElse Math.Abs(clickPoint.Y - newPosition.Y) > 15) AndAlso dragableItem IsNot Nothing
		End Function
	End Class

End Namespace
